### PR TITLE
Improve frontend lint settings

### LINKED ISSUE #469: [Conversation]: Improve Linter Settings
We should look at improving our linter to maybe get prettier to format things to our preference.

Topics I think we can talk about (feel free to add more)

---

### Unnecessary braces `{` in a directly returned function
```
const foo = () => {
  return 'bar';
};
// vs
const foo = () => 'bar';
```

--- 

### Braces around straight returns

```
const foo = (bar: string | null) => {
  if (!bar) return null;
  // vs
  if (!bar) {
    return null;
  }

  return 'baz';
};
```

---

### Auto sort imports

Get them so all our 3rd party libs are at the top

`..` sorting so that the items further away go under 3rd party imports, working down to local files `./some-file`

---

### Support more tsconfig paths

Have it so alias paths can snap us back to the root to grab types/constants/utils/services/etc

```
import { } from '../../../utilities/useSomeHook';
// vs
import { } from '~/utilities/useSomeHook';
```

Note: When it's 1 file -- it can get confusing. `import { } from 'k8sTypes'` looks weird. We need a leading character to signify it's a special import.

---

### Reassigning variables passed to you

https://github.com/opendatahub-io/odh-dashboard/pull/693#discussion_r1024165385

---

### Stop console logs of any kind

We can ignore the ones we need for stability/error reporting -- but we should stop all logs from accidentally allowing debug code from slipping in.

---

### No else return

Prevent `if () { return } else { return }` so that the code structure is more malleable in the future. 

https://eslint.org/docs/latest/rules/no-else-return

---

### Always type argumentless variables

```
let icon;
```
What is icon? Is icon the string? the JSX? We should force typing here so it has more contextual usage.

```
let icon: React.ReactNode;
```

### PR DESCRIPTION
**Pull request by DaoDaoNoCode**
## Description

closes frontend part of #469 

Frontend:
1. Use `~/...` instead of relative path
2. Use return as needed (get rid of the return statement when there is only the return in the function body)
3. Require to use curly for all the statements (if, while, etc...)
4. Encourage to name variables in camelCase
5. Prevent else return and absolute import from the src folder like `services/**`, `pages/**`, etc...
6. Require to read all the methods in the `~/api` folder not to use the specific file path
7. Add no-console rule
8. Add order rules for imports

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: frontend/src/utilities/allSettledPromises.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 type TypedPromiseRejectedResult<R> = PromiseRejectedResult & { reason: R };
 
-export const allSettledPromises = <T, E extends unknown = undefined>(
+export const allSettledPromises = <T, E = undefined>(
```
Comment (andrewballantyne) [inline frontend/src/utilities/allSettledPromises.ts:3]: Interesting... not sure why I did that originally.

**Review by andrewballantyne — COMMENTED** [review]
@DaoDaoNoCode is there a reason we are not excluding `from 'directory/file'` now that we have `~`? Eg frontend/src/services/notebookService.ts has `import { RecursivePartial } from 'typeHelpers';` and you passed the linter.

**Comment by DaoDaoNoCode** [comment]
@andrewballantyne I thought of that, and at first, I tried to create a glob pattern matching like `*/**`, but it will filter the imported from packages like `react-dom/client`. There are actually two ways, and I don't know which one is better.
1. block all the path like `*/**` and create an allowlist for the paths like `~/**`, `./**` and node packages we are using
2. create a blocklist for the paths like `services/**`, `types`, `pages/**`, and don't block any other paths

Which option do you think is better?

**Comment by andrewballantyne** [comment]
> @andrewballantyne I thought of that, and at first, I tried to create a glob pattern matching like `*/**`, but it will filter the imported from packages like `react-dom/client`. There are actually two ways, and I don't know which one is better.
> 
> 1. block all the path like `*/**` and create an allowlist for the paths like `~/**`, `./**` and node packages we are using
> 2. create a blocklist for the paths like `services/**`, `types`, `pages/**`, and don't block any other paths
> 
> Which option do you think is better?

Hmmmm -- are you sure there is no way to just block non npm dependencies? Surely there is a webpack setting or something; I swear I recall this being a thing 🤔 I'd be shocked if there was no way to stop root-level imports like `from 'directory'` because it _should_ say "that's not an npm import", just thinking about it.

**Comment by Gkrumbach07** [comment]
Is there a reason to use the tilde `~/...`.  We can get the same effect of accessing root level folders by just omitting `~/`

Example - these have the same outcome:
```ts
import { ... } from "folder"
```
```ts
import { ... } from "~/folder"

**Comment by andrewballantyne** [comment]
> Is there a reason to use the tilde `~/...`. We can get the same effect of accessing root level folders by just omitting `~/`
> 
> Example - these have the same outcome assuming folder structure has `folder` at the root level.
> 
> ```ts
> import { ... } from "folder"
> ```
> 
> ```ts
> import { ... } from "~/folder"
> ```

Yuppers -- because `import { ... } from 'thing'` is an npm import. Overloading that has no value. Putting `~` prefix indicates that it is not an npm package.

Will we conflict with a folder and a npm package? Probably not... but it is the same as "why use `;` at the end of every line". There are things we do for ease of use, common acceptance for new comers, and in general just a style we choose. I don't see any value in compounding concepts for ease of use.

Is there a reason you're against it @Gkrumbach07 ?

**Comment by Gkrumbach07** [comment]
> Is there a reason you're against it @Gkrumbach07 ?

Nope, it just seemed like extra chars at first glance. But the reasons to use it make sense

File: frontend/src/app/useTimeBasedRefresh.ts [hunk 2]
```diff
@@ -39,6 +40,7 @@ const useTimeBasedRefresh = (): SetTime => {
         `We should have refreshed but it appears the last time we auto-refreshed was less than an hour ago. '${KEY_NAME}' session storage setting can be cleared for this to refresh again within the hour from the last refresh.`,
       );
     }
+    /* eslint-enable no-console */
```
Comment (andrewballantyne) [inline frontend/src/app/useTimeBasedRefresh.ts:43]: Hmmm... wonder if we can disable block disable/enables -- seems easy to abuse.

**Review by andrewballantyne — APPROVED** [review]
Approving and merging to avoid the expected conflicts that will happen when ever single PR merges before this -- easier for them to adjust than for you to adjust.

One note I found with no-console... definitely can be a follow up though.

**Pull request merged**